src/main/scala/Core/control.scala

```diff
@@ -322,23 +336,88 @@ class Control(implicit val conf: FlexpretConfiguration) extends Module
   //val exe_brjmp = exe_valid && (exe_reg_jump || (exe_reg_branch && io.exe_br_cond))
 
   // Keep track of delay_until instruction.
+  val exe_expire_du_wu = io.exe_expire_du || io.exe_expire_wu
+  val exe_expire_ie_ee = io.exe_expire_ie || io.exe_expire_ee
+  val exe_expire = exe_expire_du_wu || exe_expire_ie_ee
+  
+  // At the end of DU, another DU instruction will be fetched because DU is a
+  // branch to self. We need to make sure we do not execute this DU, otherwise we
+  // will end in a loop.
+  // We do this by capturing the thread id that fecthed a DU, and marking it
+  // invalid when it otherwise would be executed.
```

Comment:
I do not understand this. DU should not "branch" to itself when the DU time has already passed. The argument is an absolute time-point. So when it wakes up, fetches DU again, we have passed the absolute time and it should just continue?

Reply:
What I wrote is based on waveform observations - I saw that for some HW thread configurations (1, 2, 3 threads), the thread executing a DU would fetch another DU when the timer had expired and execute it. Then it would just constantly loop between waking up from DU, fetching another DU, executing it, waking up, ...

The signal that indicates whether a DU has expired is called `du_expired`. This signal is per-thread (which is necessary for handling multi-threaded DUs), so it will be high if
1. thread A has executed a DU,
2. the DU has expired,
3. and thread A currently is inside the execute stage. (The time comparision happens inside the CSR module, and the CSR module is given the thread ID of the exe stage.)

The last condition is what makes it difficult. Because for, say, HWTHREADS=2, the `du_expired` will not be high before thread A has already fetched another DU (analogous to branching, it does not know the DU is the wrong instruction to fetch, because the result is not ready yet). 

For this reason, I think some kind of  "squasing/kill" mechanism for DU instructions needs to be in place. However, it is not obvious to determine which instruction to kill either. The code added is an implementation of doing this correctly for any configuration of HWTHREADS and also any schedule (in the flex case).

I have tested these mechanisms quite a lot as well for all HWTHREAD configurations. See `test_fp_wait_until`, `test_fp_delay_until`, `test_du_not_stopped_by_int`, `test_wu_stopped_by_int`. https://github.com/pretis/flexpret/blob/context-switch/programs/tests/c-tests/interrupt/interrupt.c

Reply:
Aha, I see. So there was actually a CPU bug here that you are know fixing.

Reply:
DU/WU was not implemented for several hardware threads, so it's more of a missing feature. There used to be only one register that contained the expire time, so if several threads tried to run DU/WU at the same time, they would override each other's expire time. So this is 'unexplored' waters... But from extensive testing I think it works quite well now.